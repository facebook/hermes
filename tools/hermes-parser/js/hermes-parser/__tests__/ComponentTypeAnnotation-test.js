/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow strict-local
 * @format
 */

import {
  printForSnapshotESTree,
  parseForSnapshotESTree,
  printForSnapshotBabel,
  parseForSnapshotBabel,
} from '../__test_utils__/parse';

describe('ComponentTypeAnnotation', () => {
  describe('Basic', () => {
    const code = `
      type T = component();
    `;

    test('ESTree', async () => {
      expect(await parseForSnapshotESTree(code)).toMatchSnapshot();
      expect(await printForSnapshotESTree(code)).toBe(code.trim());
    });

    test('Babel', async () => {
      expect(await parseForSnapshotBabel(code)).toMatchSnapshot();
      expect(await printForSnapshotBabel(code)).toMatchInlineSnapshot(
        `"type T = any;"`,
      );
    });
  });

  describe('Union', () => {
    const code = `
type T = component() | null;
type T = (component() renders Foo) | null;
    `;

    test('ESTree', async () => {
      expect(await parseForSnapshotESTree(code)).toMatchSnapshot();
      expect(await printForSnapshotESTree(code)).toBe(code.trim());
    });

    test('Babel', async () => {
      expect(await parseForSnapshotBabel(code)).toMatchSnapshot();
      expect(await printForSnapshotBabel(code)).toMatchInlineSnapshot(`
        "type T = any | null;
        type T = any | null;"
      `);
    });
  });
  describe('Without parens union', () => {
    const code = `
      type T = component() renders Foo | null;
    `;

    test('ESTree', async () => {
      expect(await printForSnapshotESTree(code)).toBe(
        `type T = (component() renders Foo) | null;`,
      );
    });
  });
});
