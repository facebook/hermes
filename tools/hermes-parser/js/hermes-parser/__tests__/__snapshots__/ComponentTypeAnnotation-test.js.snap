// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ComponentTypeAnnotation Basic Babel 1`] = `
{
  "body": [
    {
      "id": {
        "name": "T",
        "type": "Identifier",
      },
      "right": {
        "type": "AnyTypeAnnotation",
      },
      "type": "TypeAlias",
      "typeParameters": null,
    },
  ],
  "type": "Program",
}
`;

exports[`ComponentTypeAnnotation Basic ESTree 1`] = `
{
  "body": [
    {
      "id": {
        "name": "T",
        "optional": false,
        "type": "Identifier",
        "typeAnnotation": null,
      },
      "right": {
        "params": [],
        "rendersType": null,
        "rest": null,
        "type": "ComponentTypeAnnotation",
        "typeParameters": null,
      },
      "type": "TypeAlias",
      "typeParameters": null,
    },
  ],
  "type": "Program",
}
`;

exports[`ComponentTypeAnnotation Union Babel 1`] = `
{
  "body": [
    {
      "id": {
        "name": "T",
        "type": "Identifier",
      },
      "right": {
        "type": "UnionTypeAnnotation",
        "types": [
          {
            "type": "AnyTypeAnnotation",
          },
          {
            "type": "NullLiteralTypeAnnotation",
          },
        ],
      },
      "type": "TypeAlias",
      "typeParameters": null,
    },
    {
      "id": {
        "name": "T",
        "type": "Identifier",
      },
      "right": {
        "type": "UnionTypeAnnotation",
        "types": [
          {
            "type": "AnyTypeAnnotation",
          },
          {
            "type": "NullLiteralTypeAnnotation",
          },
        ],
      },
      "type": "TypeAlias",
      "typeParameters": null,
    },
  ],
  "type": "Program",
}
`;

exports[`ComponentTypeAnnotation Union ESTree 1`] = `
{
  "body": [
    {
      "id": {
        "name": "T",
        "optional": false,
        "type": "Identifier",
        "typeAnnotation": null,
      },
      "right": {
        "type": "UnionTypeAnnotation",
        "types": [
          {
            "params": [],
            "rendersType": null,
            "rest": null,
            "type": "ComponentTypeAnnotation",
            "typeParameters": null,
          },
          {
            "type": "NullLiteralTypeAnnotation",
          },
        ],
      },
      "type": "TypeAlias",
      "typeParameters": null,
    },
    {
      "id": {
        "name": "T",
        "optional": false,
        "type": "Identifier",
        "typeAnnotation": null,
      },
      "right": {
        "type": "UnionTypeAnnotation",
        "types": [
          {
            "params": [],
            "rendersType": {
              "operator": "renders",
              "type": "TypeOperator",
              "typeAnnotation": {
                "id": {
                  "name": "Foo",
                  "optional": false,
                  "type": "Identifier",
                  "typeAnnotation": null,
                },
                "type": "GenericTypeAnnotation",
                "typeParameters": null,
              },
            },
            "rest": null,
            "type": "ComponentTypeAnnotation",
            "typeParameters": null,
          },
          {
            "type": "NullLiteralTypeAnnotation",
          },
        ],
      },
      "type": "TypeAlias",
      "typeParameters": null,
    },
  ],
  "type": "Program",
}
`;
